src/RFC_0388_BearerTokens.md

```diff
@@ -0,0 +1,206 @@
+that a user owns on their behalf. Ethereum and many other blockchains achieve this in a stateful
+manner by invoking methods such as 'approve' and 'approve_all' in ERC20 and ERC721. Being stateful, 
+the user must spend fees in order to add or revoke permissions. When fees are high, or due to 
+simplistic user interfaces, a user will often grant a much higher level of permission than is required. 
+For example, the user may approve a service to transfer all of their funds, or all of their NFTs.
+
+The scheme proposed by this RFC is a stateless token that the bearer can use to invoke methods on the 
+DAN layer.
+
+## Related Requests for Comment
+
+## Description
+
+This scheme is inspired by [Macaroons] in that it allows the bearer of a token to delegate a more restrictive token to
+another bearer. 
+
+
+## Structure
+
+### Auth Token
+
+| Field | Type | Description |
+| --- | --- | --- |
+| root_nonce | bytes(32) (optional) | a base ID used to revoke permissions(see below)  | 
+| expires_at_height | u64 (optional) | the sidechain height at which this token expires. Timestamps are not reliable in blockchains, so height is used in this case |
+| granted_to | pubkey | the public key of the grantee. Any bearer using this token will need to prove knowledge of the private key |
+| scopes | string[] | A list of scopes granted to this scope |
+| caveats | CaveatExpression[] | An ordered list of caveats |
+| based_on | Token (optional) | If this token is derived from another token, it should be present here |
+| issuer | PubKey | the issuer of this token |
+| issuer_sig | Signature | a signature of the challenge `Hash(base_id + granted_to + scopes + caveats + expires_at_height + based_on + issuer )` |
+
+### Caveat Expression
+
+``` 
+Caveat Expression = <Field> <operator> <Argument>
+```
+
+Field: An arbitrary string that will be interpreted by the code
+Operator: OneOf("eq", "le", "lt","ge", "gt")
+Argument: A constant value, to be interpretted by the code
+
+Examples
+```
+amount lt 1000
+token_id eq 4759
+```
+
+## Delegation
+
+A bearer of a token may grant another identity a more specific token, provided that the scopes and caveats are more restrictive.
+
+
+## Validation
+
+The root_nonce, if specified must match the root nonce on record for the issuer.
+> Open question: should root nonce just be a special caveat?
+
+Before starting execution of the instruction, the list of auth tokens must be validated to ensure that each token is more restrictive than
+the last.
+
+When executing instructions, caveats MUST be checked if they are relevant to the resources being acted upon. 
+For example, if a function requires a scope `transfer`, the most specific AuthToken must have that scope present.
+
+
+
+## Revoking Tokens
+Each contract SHOULD store a root nonce for each identity in the contract. To revoke a set of tokens, an identity owner
+may change their root nonce. This will revoke all tokens based on this root.
+
```

Comment:
To clearly distinguish between asset tokens in the contract and authority bearer tokens, let's use a different word than "token" for the auth tokens.

Options could be:
- meringues (similar to macaroons)
- badges (implies authority)
- permits
- mandate
- accreditation
-